### PR TITLE
Block Patterns: Load patterns from wordpress.org API

### PR DESCRIPTION
**Pull request by ryelle**
## Description
As part of the WordPress.org pattern directory project, we want to load patterns remotely from WordPress.org — this PR is a first step to get there. It unregisters & removes the existing core patterns, and pulls down the patterns from an API on [wordpress.org/patterns](https://wordpress.org/patterns/). These are the same patterns that were just added in #29973. With this change, patterns can be added or updated without needing to release new WordPress or Gutenberg versions.

Right now this preloads the patterns server-side, so functionally there should be no change to the editing experience. Eventually we'll want to load dynamically, client-side, so that authors can search & filter through all patterns on wordpress.org/patterns. See https://github.com/WordPress/pattern-directory/issues/10

## How has this been tested?

- Add or edit a post
- Open the inserter and look at the non-theme patterns
- They should exist 🙂 
- Try to add a core pattern (ex: "Three columns with offset images")
- It should add correctly

~Note: The patterns are rendered as a classic blocks for now, that will be fixed with #28799 - leaving this PR in draft until then.~ This was fixed.

Note 2: There are two patterns - heading and quote - which currently add `blockTypes` support for the heading and quote blocks, respectively. we don't have a way of flagging that on the pattern directory yet, so they're registered without it. See https://github.com/WordPress/pattern-directory/issues/33#issuecomment-824317557

### CODE REVIEW

File: lib/block-patterns.php
Comment (ryelle) [inline lib/block-patterns.php]: I don't think this should change, but if I leave it as `init`, I get this error when running the unit tests:

> Fatal error: Uncaught Error: Class 'Spy_REST_Server' not found in /var/www/html/wp-includes/rest-api.php:509

This was happening locally and in [the PR's unit test action](https://github.com/WordPress/gutenberg/runs/2403775789).

File: lib/block-patterns.php
Comment (StevenDufresne) [inline lib/block-patterns.php]: Just noting that without the comment this would be impossible to understand since the keyword ID lives outside of this project. Additionally, if any changes happen in the directory, this would return incorrect patterns. I think for now it's ok, but if core patterns maintain priority in the future, we may want to consider a different approach to loading core patterns that doesn't rely on a hard coded id.

**Review by StevenDufresne — COMMENTED** [review]
Great. This seems as simple as it looks. 

Is there any value in leaving in the `$new_core_block_patterns` and merge in new ones from the API as a fallback to limit the scope of this change as we work on the directory?

File: lib/block-patterns.php
Comment (ryelle) [reply lib/block-patterns.php]: That's why I left the comment 🙂 

This really is a "for now" kind of feature — to move the management of these core patterns into the block pattern directory. The longer-term plan is for a more dynamic search UI that fetches from the API on the client side (like the block directory), any prioritization would happen on the API side, so Gutenberg wouldn't need to be aware of this category at all.

File: lib/block-patterns.php
Comment (andrewserong) [reply lib/block-patterns.php]: Would it be worth using a more restrictive hook so that the API call isn't being made on admin screens that don't use patterns? In WordPress.com, where we use [a similar approach](https://github.com/Automattic/wp-calypso/blob/94c1213c164e6e2e6f9d5ecd19ff769a5688303c/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L247) to load patterns from an API endpoint we decided to use the `current_screen` hook and check for `is_block_editor` or `gutenberg_is_edit_site_page()` so that we only call the endpoint on views that need it.

File: lib/block-patterns.php
Comment (andrewserong) [inline lib/block-patterns.php:215]: Another issue we ran into over in WordPress.com was calling the endpoint on every request for the admin page. We decided to [guard the call](https://github.com/Automattic/wp-calypso/blob/94c1213c164e6e2e6f9d5ecd19ff769a5688303c/apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php#L175) behind a `wp_cache_get` check and then `wp_cache_set` the results from the API endpoint for a period of time so that each admin page request doesn't fire off an external API call.

**Review by andrewserong — COMMENTED** [review]
Hi there, I hope you don't mind the drive-by commenting, but I thought I'd add a couple of tiny comments based on some experiences building a [similar feature](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php) into a plugin in WordPress.com, that I thought could be relevant. Feel to disregard, of course! And I'm very excited and curious to see how the patterns directory integration progresses 🙂

File: lib/class-wp-rest-pattern-directory-controller.php
Comment (dd32) [inline lib/class-wp-rest-pattern-directory-controller.php]: Since Gutenberg is the primary consumer of the api.wordpress.org/patterns/ API, could this be made the default of the API instead of passing it?

**Comment by StevenDufresne** [comment]
@iandunn @dd32 It may be an obvious question but are there any last-minute hesitations on the API side that need to be considered?

File: lib/block-patterns.php
Comment (dd32) [reply lib/block-patterns.php:215]: I would also recommend using some caching around `rest_do_request()` in this case, a transient makes sense to me. I don't consider this to be a blocker for merge though.

**Comment by dd32** [comment]
@StevenDufresne The only major concern regarding the API is to ensure that we have proper caching on the endpoints (both on WordPress.org, and within Core for anything that is called on every editor load) before the Gutenberg release is included in a WordPress release.

I've left some comments on a few things I'd like to change, but they're not blockers for merge, but would preferably be taken into account.

**Comment by draganescu** [comment]
I tested this and it does not seem to break or make any difference other than for me the patterns' previews loaded slower.

**Comment by priethor** [comment]
The pattern previews loaded slower for me as well (although I understand they shouldn't as patterns are preloaded server-side), which sometimes results in a visual glitch when there are many patterns in a category and once the previews finish loading:

https://user-images.githubusercontent.com/27339341/118647295-785fc900-b7e1-11eb-8fa7-4b2e8b5de660.mov

I can't reproduce it consistently on local environments, but on Gutenberg.run it happens every time, both in Chrome and Firefox.

File: lib/block-patterns.php
Comment (draganescu) [reply lib/block-patterns.php]: @andrewserong has a good point on employing a more restrictive hook, Before we used an API call it didn't matter, but now it does. Also, we're prefetching a lot of data in the Gutenberg editors. Shouldn't this be treated the same way?

File: lib/block-patterns.php
Comment (draganescu) [reply lib/block-patterns.php]: Can we make that `11` be a variable with a good name like `$core_keyword_id = 11` and also have that comment?

Comment (draganescu) [reply lib/block-patterns.php]: How come the param named `keyword` is a number? I would expect a string for keyword ...

**Comment by draganescu** [comment]
@youknowriad if this is merged as a Gutenberg experiment until it gets fixed, can it still land in 5.8? Or does it have to be merged as a feature now?

**Comment by youknowriad** [comment]
Historically, features need to land as stable and in Core (not just the plugin) before the feature freeze to make it to the WordPress release. While I'm really excited about this feature, late features always make me nervous so I'm wondering whether it's wiser to defer until 5.9 to give it time to mature. How confident are you all about its current state?

File: lib/block-patterns.php
Comment (TimothyBJacobs) [reply lib/block-patterns.php:215]: Should we do the caching in the actual endpoint around the API call to pattern directory? I think we'd want to avoid this returning different data from the endpoint?

File: lib/block-patterns.php
Comment (ryelle) [reply lib/block-patterns.php]: This data isn't used the same way as the other prefetched data - it's not used in gutenberg directly, it's only used on the server-side to register patterns. Once the pattern directory is more dynamic, and block patterns are registered in JS, we can switch to something more like the other prefetched endpoints.

In 89045a6b3755c80f978f8c7329b0b18b6267f921, I updated the code here to use the `current_screen` approach for loading pattern content, but left the unregistration on `init`, to match what's happening in core now.

File: lib/block-patterns.php
Comment (ryelle) [reply lib/block-patterns.php]: It's a taxonomy on wp.org/patterns. Is that a problem? I was trying to make the least changes to the wp.org api, but this could probably be updated if so. cc @iandunn

File: lib/class-wp-rest-pattern-directory-controller.php
Comment (ryelle) [reply lib/class-wp-rest-pattern-directory-controller.php]: Removed - f57bafa6e26192a10e889b201e7dc537cdafc756

I haven't updated the wporg api yet.

**Comment by ryelle** [comment]
> The only major concern regarding the API is to ensure that we have proper caching on the endpoints (both on WordPress.org, and within Core for anything that is called on every editor load) before the Gutenberg release is included in a WordPress release.

Caching still needs to be enabled on wp.org (see https://github.com/WordPress/pattern-directory/issues/99). I added an hour transient cache for the pattern request from the editor, will that work?

> The pattern previews loaded slower for me as well (although I understand they shouldn't as patterns are preloaded server-side), which sometimes results in a visual glitch when there are many patterns in a category and once the previews finish loading

Does that happen on trunk? This PR isn't touching any of the query blocks, so I'm not sure why that would happen here. They also shouldn't be loading any slower… the only change here is server-side, it shouldn't affect the loading once the patterns are set into `__experimentalBlockPatterns`. I've switched back and forth between the branch & trunk, and I don't see a significant difference.

> While I'm really excited about this feature, late features always make me nervous so I'm wondering whether it's wiser to defer until 5.9 to give it time to mature. How confident are you all about its current state?

This was intended to lay some groundwork for more dynamic block patterns, and make it easier to manage the core patterns. There shouldn't be any functionality change in gutenberg, it _should_ just swap out the pattern files for the same data from an API. I think the only outlying issue is the perceived slowness. If @draganescu or @priethor could re-test, maybe the issue was how out-of-date the branch was.

File: lib/block-patterns.php
Comment (iandunn) [reply lib/block-patterns.php]: It's been awhile, but if I'm reading everything correctly, I think we could rename the `keyword` param to `keyword_id` (or maybe plural), and it'd only effect `WP_REST_Pattern_Directory_Controller::get_items()`.

On the API side it's `pattern-categories`, and `get_items()` maps between the two.

Using a slug would be nice, but IIRC the REST API only accepts IDs, unless we add custom code. I haven't checked that, though.

Either way, IMO it's best to be practical in this case. As long as variable names and other easily updated things are clear, I think that's good enough. No need to get bogged down over something small.

**Comment by tellyworth** [comment]
> How confident are you all about its current state?

The API side is solid enough. For this PR it only ever needs to return the core patterns, which it does already. That can easily be cached (or even made static) on the server side for performance.

Getting this in to 5.8 means two things: core patterns can then be improved and updated independent of the Gutenberg and WP release cycles. And, since this PR contains the basic core of the necessary editor-side code, adding support for third-party patterns in 5.9 will become easier. 

Defering this patch to 5.9 means that wide testing of remotely-fetched core patterns doesn't start until the 5.9 beta cycle, and that probably pushes work on third-party patterns to the next release cycle after that.

**Review by draganescu — APPROVED** [review]
Yes, tested again and the perceived speed is greatly improved. Given that it does minimal changes and the we can improve caching in subsequent PRs.

**Comment by youknowriad** [comment]
@ryelle @tellyworth sounds good to me, I'll defer this to you all on this.

**Comment by youknowriad** [comment]
Does it mean we need a follow-up trac ticket to remove the bundled patterns from there as well? (probably before next Tuesday)

**Comment by youknowriad** [comment]
Also with this in place, we can still remove support for the core patterns right?

**Comment by ryelle** [comment]
> Does it mean we need a follow-up trac ticket to remove the bundled patterns from there as well?

If they won't be removed as part of the gutenberg update/merge, then yes.

> Also with this in place, we can still remove support for the core patterns right?

Do you mean like #31922? We could still add a filter to short-circuit registration, like #31924 does.

**Comment by youknowriad** [comment]
I meant removing the bundled core block patterns by doing

```
remove_theme_support( 'core-block-patterns' );
```

like shown on that post https://make.wordpress.org/core/2020/07/16/block-patterns-in-wordpress-5-5/

**Comment by ryelle** [comment]
Ah, I see. Now it does - if no theme support, these patterns won't be registered.

**Comment by youknowriad** [comment]
Thanks for the update about the theme support flag.

> If they won't be removed as part of the gutenberg update/merge, then yes.

PHP changes don't get back ported automatically, so we have two options there:

 - Including the required php changes in the package update patch if these changes are mandatory for the JS packages update
 - If it can be a separate change, a dedicate ticket might be needed.

For this particular PR, I'm not sure. Basically the question is what happens if we update the JS packages with the changes of this PR without the php changes.

File: lib/block-patterns.php
Comment (mcsf) [inline lib/block-patterns.php:216]: Don't let this block this PR, but: how can we make this value less magic? Could this option exist more explicitly in the parent API (`/__experimental/pattern-directory/patterns`) rather than just guessing a term identifier?

**Comment by ryelle** [comment]
> Basically the question is what happens if we update the JS packages with the changes of this PR without the php changes.

Nothing, since there are no JS changes here.

File: lib/block-patterns.php
Comment (iandunn) [reply lib/block-patterns.php:216]: There's a discussion about that at https://github.com/WordPress/gutenberg/pull/28800#discussion_r617986203

**Comment by youknowriad** [comment]
The e2e failure here sounds legit. (adding patterns). Anyone to take a look?

**Comment by ryelle** [comment]
yeah, I can fix it.

**Comment by noisysocks** [comment]
I rebased this to see if it fixes E2E failures.

**Review by noisysocks — APPROVED** [review]
Works when I test locally! 👍

**Comment by noisysocks** [comment]
I just noticed this introduces some PHP warnings which appear on pages without a block editor. I've fixed them in https://github.com/WordPress/gutenberg/pull/32025.

**Pull request merged**
